src/components/app/index.tsx

```diff
@@ -25,33 +66,131 @@ export default class App extends Component<Props, State> {
+  optionsUpdated() {
+    if (!this.optionsUpdateTimer) {
+      this.optionsUpdateTimer = setTimeout(this.commitOptionsUpdated, 500);
+    }
+  }
+
+  @bind
+  commitOptionsUpdated() {
+    this.optionsUpdateTimer = null;
+    this.updateImages();
+  }
+
   @bind
   async onFileChange(event: Event) {
+    this.setState({ loading: true });
     const fileInput = event.target as HTMLInputElement;
     if (!fileInput.files || !fileInput.files[0]) return;
     // TODO: handle decode error
-    const bitmap = await createImageBitmap(fileInput.files[0]);
-    const data = await bitmapToImageData(bitmap);
-    const encoder = new MozJpegEncoder();
-    const compressedData = await encoder.encode(data);
-    const blob = new Blob([compressedData], {type: 'image/jpeg'});
-    const compressedImage = await createImageBitmap(blob);
-    this.setState({ img: compressedImage });
+    const sourceImg = await createImageBitmap(fileInput.files[0]);
+    // compute the corresponding ImageData once since it only changes when the file changes:
+    const sourceData = await bitmapToImageData(sourceImg);
+    this.setState({ sourceImg, sourceData, loading: false });
+    this.updateImages();
   }
 
-  render({ }: Props, { img }: State) {
+  async updateImages() {
+    const { sourceData, leftType, rightType, leftOptions, rightOptions } = this.state;
+    if (!sourceData) return;
+    const id = ++this.compressCounter;
+    this.setState({ loading: true });
+    const [leftImg, rightImg] = await Promise.all([
+      this.updateCompressedImage(sourceData, leftType, leftOptions),
+      this.updateCompressedImage(sourceData, rightType, rightOptions)
+    ]);
+    if (this.compressCounter !== id) return;
+    this.setState({ leftImg, rightImg, loading: false });
+  }
+
+  async updateCompressedImage(sourceData: ImageData, type: ImageType, options: CodecOptions, retries= 0) {
+    if (type === 'original') {
+      return this.state.sourceImg;
+    }
+    // @todo reuse here crashes
+    // let encoder = this.encoders[type];
+    // if (!encoder) {
+    //   encoder = this.encoders[type] = new AllEncoders[type]();
+    // }
+    // if (!encoder) {
+    //   console.error(`Unknown encoder: ${type}`);
+    //   return;
+    // }
+    const encoder = new AllEncoders[type]();
+    try {
+      const compressedData = await encoder.encode(sourceData, options);
+      const blob = new Blob([compressedData], { type: 'image/jpeg' });
+      return await createImageBitmap(blob);
+    } catch (err) {
+      // console.log(`failed to encode ${type} (retries: ${retries}): ${err}`);
+      if (retries < 5) {
+        await this.updateCompressedImage(sourceData, type, options, retries + 1);
+      }
+    }
+  }
+
+  render({ }: Props, { loading, leftType, leftOptions, rightType, rightOptions, leftImg, rightImg }: State) {
     return (
       <div id="app" class={style.app}>
-        {img ?
-          <Output img={img} />
-          :
-          <div>
+        {leftImg && rightImg ? (
```

Comment:
Eventually we'll go to the output view even if only one of these are ready. Not needed for this PR.